src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -398,16 +420,92 @@ Status GcsActorManager::RegisterActor(
+        // requests are included) and remove all of them from
+        // actor_to_register_callbacks_.
+        // Reply to the owner to indicate that the actor has been registered.
+        auto iter = actor_to_register_callbacks_.find(actor->GetActorID());
+        RAY_CHECK(iter != actor_to_register_callbacks_.end() && !iter->second.empty());
+        for (auto &callback : iter->second) {
+          callback(actor);
+        }
+        actor_to_register_callbacks_.erase(iter);
+      }));
+  return Status::OK();
+}
+
+Status GcsActorManager::ReportActorDependenciesResolved(
+    const ray::rpc::ReportActorDependenciesResolvedRequest &request,
+    ReportActorDependenciesResolvedCallback callback) {
+  RAY_CHECK(callback);
+  const auto &actor_creation_task_spec = request.task_spec().actor_creation_task_spec();
+  auto actor_id = ActorID::FromBinary(actor_creation_task_spec.actor_id());
+
+  auto iter = registered_actors_.find(actor_id);
+  if (iter != registered_actors_.end() &&
+      iter->second->GetState() == rpc::ActorTableData::ALIVE) {
+    // When the network fails, Driver/Worker is not sure whether GcsServer has received
+    // the request, so Driver/Worker will try again and again until receiving the reply
+    // from GcsServer. If the actor has been created successfully then just reply to the
+    // caller.
+    callback(iter->second);
+    return Status::OK();
+  }
+
+  auto pending_report_iter = actor_to_report_callbacks_.find(actor_id);
+  if (pending_report_iter != actor_to_report_callbacks_.end()) {
+    // It is a duplicate message, just mark the callback as pending and invoke it after
+    // the actor has been successfully created.
+    pending_report_iter->second.emplace_back(std::move(callback));
+    return Status::OK();
+  }
+  // Mark the callback as pending and invoke it after the actor has been successfully
+  // created.
+  actor_to_report_callbacks_[actor_id].emplace_back(std::move(callback));
+
+  // Remove the actor from the unresolved actor map.
+  auto actor = std::make_shared<GcsActor>(request.task_spec());
+  actor->GetMutableActorTableData()->set_state(rpc::ActorTableData::PENDING_CREATION);
+  const auto &owner_address = actor->GetOwnerAddress();
+  auto node_id = ClientID::FromBinary(owner_address.raylet_id());
+  auto worker_id = WorkerID::FromBinary(owner_address.worker_id());
+  auto it = unresolved_actors_.find(node_id);
+  RAY_CHECK(it != unresolved_actors_.end());
+  auto worker_to_actors_it = it->second.find(worker_id);
+  RAY_CHECK(worker_to_actors_it != it->second.end());
+  RAY_CHECK(worker_to_actors_it->second.erase(actor_id) != 0);
+  if (worker_to_actors_it->second.empty()) {
+    it->second.erase(worker_to_actors_it);
+    if (it->second.empty()) {
+      unresolved_actors_.erase(it);
+    }
+  }
+  // Update the resolved actor to the `registered_actors_` as the task specification is
+  // changed.
+  registered_actors_[actor_id] = actor;
```

Comment:
It'd be clearer to put the actor in `registered_actors_` during `RegisterActor`. So we can consider `registered_actors_` as all actors.

Reply:
> It'd be clearer to put the actor in `registered_actors_` during `RegisterActor`. So we can consider `registered_actors_` as all actors.

It is already put into `registered_actors_` when `RegisterActor`.